File: recipes/perl/storage/replication_target_on_dv_operations_end_to_end.pl
#!/usr/bin/env perl

use LWP::UserAgent;
use LWP::Protocol::https;
print "LWP::UserAgent: ".LWP::UserAgent->VERSION,"\n";
print "LWP::Protocol::https: ".LWP::Protocol::https->VERSION,"\n";
use JSON;
use Getopt::Long qw(GetOptions);
use storage;

# This script consists of the helper functions to excute NetBackup APIs to create storage unit.
# 1) Login to Netbackup
# 2) Create storage server
# 3) Create disk Pool
# 4) Create storage unit

#
# The token is the key to the NetBackup AuthN/AuthZ scheme.  You must login and get a token
# and use this token in your Authorization header for all subsequent requests.  Token validity
# is fixed at 24 hours
#
my $token;

my $protocol = "https";
my $port = "1556";
my $nbmaster;
my $username;
my $password;
my $sts_payload;
my $dp_payload;
my $stu_payload;
my $domainName;
my $domainType;
my $base_url;


# subroutines for printing usage and library information required to run the script.
sub print_usage {
    print("\n\nUsage:");
    print("\nperl replication_target_operations_end_to_end.pl -nbmaster <master_server> -username <username> -password <password> -sts_payload <input JSON for storage server> -dp_payload <input JSON for disk pool> -stu_payload <input JSON for storage unit> -add_reptarget_payload <input JSON for add replication target> -delete_reptarget_payload <input JSON for delete replication target> [-domainname <domain_name>] [-domaintype <domain_type>]\n\n\n")
}

sub print_disclaimer {
    print("--------------------------------------------------------\n");
    print("--       This script requires Perl 5.20.2 or later    --\n");
    print("--------------------------------------------------------\n");
    print("Executing this library requires some additional libraries like \n\t'LWP' \n\t'JSON'\ \n\t'Getopt'\ \n\n");
    print("You can specify the 'nbmaster', 'username', 'password', 'sts_payload', 'dp_payload', 'stu_payload', 'add_reptarget_payload', '-delete_reptarget_payload', 'domainName' and 'domainType' as command-line parameters\n");
    print_usage();
}

# subroutine to process user input
sub user_input {
    GetOptions(
        'nbmaster=s' => \$nbmaster,
        'username=s' => \$username,
        'password=s' => \$password,
        'sts_payload=s' => \$sts_payload,
        'dp_payload=s' => \$dp_payload,
        'stu_payload=s' => \$stu_payload,
        'add_reptarget_payload=s' => \$add_reptarget_payload,
        'delete_reptarget_payload=s' => \$delete_reptarget_payload,
        'domainName=s' => \$domainName,
        'domainType=s' => \$domainType,
    ) or die print_usage();

    if ($nbmaster eq "") {
        print("Please provide the value for 'nbmaster'");
        exit;
    }

    if ($username eq "") {
        print("Please provide the value for 'username'");
        exit;
    }

    if ($password eq "") {
        print("Please provide the value for 'password'");
        exit;
    }
    if ($sts_payload eq "") {
        print("Please provide the value for 'sts_payload'");
        exit;
    }

    if ($dp_payload eq "") {
        print("Please provide the value for 'dp_payload'");
        exit;
    }

    if ($stu_payload eq "") {
        print("Please provide the value for 'stu_payload'");
        exit;
    }

    if ($add_reptarget_payload eq "") {
        print("Please provide the value for 'add_reptarget_payload'");
        exit;
    }

    if ($delete_reptarget_payload eq "") {
        print("Please provide the value for 'delete_reptarget_payload'");
        exit;
    }

    $base_url = "$protocol://$nbmaster:$port/netbackup";
}

sub storage_api_automation {
    my $token = storage::perform_login($nbmaster, $username, $password, $domain_name, $domain_type);
    my $response_sts = storage::post_storage_server($nbmaster, $token, $sts_payload);
    my $response_dp = storage::post_disk_pool($nbmaster, $token, $dp_payload);
    storage::post_storage_unit($nbmaster, $token, $stu_payload);

    # Get the stsid from response_sts, dvid from response_dp.
    my $stsid = $response_sts->{data}->{id}
    my $dvid = $response_dp->{data}->{attributes}->{diskVolumes}[0]->{name}

    storage::post_add_replication_target_on_dv($nbmaster, $token, $stsid, $dvid, $add_reptarget_payload)
    my $response_reptarget = storage::get_all_replication_targets($nbmaster, $token, $stsid, $dvid)
    my $reptargetid = $response_reptarget->{data}[0]->{id}

    storage::get_replication_targets_by_id($nbmaster, $token, $stsid, $dvid, $reptargetid)

    storage::post_delete_replication_target_on_dv($nbmaster, $token, $stsid, $dvid, $delete_reptarget_payload)
}

print_disclaimer();

user_input();

storage_api_automation();

File: recipes/python/storage/replication_target_on_dv_operations_end_to_end.py
import sys
import storage
import json
import texttable as tt

# This script consists of the helper functions to excute NetBackup APIs to create storage unit.
# 1) Login to Netbackup
# 2) Create storage server
# 3) Create disk Pool
# 4) Create storage unit

protocol = "https"
nbmaster = ""
username = ""
password = ""
domainname = ""
domaintype = ""

port = 1556

def print_usage():
	print("Example:")
	print("python -W ignore replication_target_operations_end_to_end.py -nbmaster <master_server> -username <username> -password <password> -sts_payload <input JSON for storage server> -dp_payload <input JSON for disk pool> -stu_payload <input JSON for storage unit> -add_reptarget_payload <input JSON for add replication target> -delete_reptarget_payload <input JSON for delete replication target> [-domainname <domain_name>] [-domaintype <domain_type>]\n\n\n")

def read_command_line_arguments():
	if len(sys.argv)%2 == 0:
		print_usage()
		exit()

	global nbmaster
	global username
	global password
	global domainname
	global domaintype
	global sts_payload
	global dp_payload
	global stu_payload
	global add_reptarget_payload
	global delete_reptarget_payload
	global stsid
	global dvid
	global reptargetid

	for i in range(1, len(sys.argv), 2):
		if sys.argv[i] == "-nbmaster":
			nbmaster = sys.argv[i + 1]
		elif sys.argv[i] == "-username":
			username = sys.argv[i + 1]
		elif sys.argv[i] == "-password":
			password = sys.argv[i + 1]
		elif sys.argv[i] == "-sts_payload":
			sts_payload = sys.argv[i + 1]
		elif sys.argv[i] == "-dp_payload":
			dp_payload = sys.argv[i + 1]
		elif sys.argv[i] == "-stu_payload":
			stu_payload = sys.argv[i + 1]
		elif sys.argv[i] == "-add_reptarget_payload":
			add_reptarget_payload = sys.argv[i + 1]
		elif sys.argv[i] == "-delete_reptarget_payload":
			delete_reptarget_payload = sys.argv[i + 1]
		elif sys.argv[i] == "-domainname":
			domainname = sys.argv[i + 1]
		elif sys.argv[i] == "-domaintype":
			domaintype = sys.argv[i + 1]
		else:
			print_usage()
			exit()

	if nbmaster == "":
		print("Please provide the value for 'nbmaster'")
		exit()
	elif username == "":
		print("Please provide the value for 'username'")
		exit()
	elif password == "":
		print("Please provide the value for 'password'")
		exit()
	elif sts_payload == "":
		print("Please provide the value for 'sts_payload'")
		exit()
	elif dp_payload == "":
		print("Please provide the value for 'dp_payload'")
		exit()
	elif stu_payload == "":
		print("Please provide the value for 'stu_payload'")
		exit()
	elif add_reptarget_payload == "":
		print("Please provide the value for 'add_reptarget_payload'")
	elif delete_reptarget_payload == "":
		print("Please provide the value for 'delete_reptarget_payload'")


read_command_line_arguments()

base_url = protocol + "://" + nbmaster + ":" + str(port) + "/netbackup"

jwt = storage.perform_login(username, password, base_url, domainname, domaintype)

response_sts = storage.create_storage_server(jwt, base_url, sts_payload)
print(response_sts)

response_dp = storage.create_disk_pool(jwt, base_url, dp_payload)
print(response_dp)

response_stu = storage.create_storage_unit(jwt, base_url, stu_payload)
print(response_stu)

# Get the stsid from response_sts, dvid from response_dp.

stsid = response_sts['data']['id']
dvid = response_dp['data']['attributes']['diskVolumes'][0]['name']

storage.add_replication_target_on_diskvolume(jwt, base_url, add_reptarget_payload, stsid, dvid)

response_reptarget = storage.get_all_replication_targets_on_diskvolume(jwt, base_url, stsid, dvid)
print(response_reptarget)

# get reptargetid from response_reptarget
reptargetid = response_reptarget['data'][0]['id']
response_reptarget = storage.get_replication_target_by_id_on_diskvolume(jwt, base_url, stsid, dvid, reptargetid)
print(response_reptarget)

storage.delete_replication_target_on_diskvolume(jwt, base_url, delete_reptarget_payload, stsid, dvid)

